aws/resource_aws_batch_job_queue.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{batch.JQStatusCreating, batch.JQStatusUpdating},
+		Target:     []string{batch.JQStatusValid},
+		Refresh:    batchJobQueueRefreshFunc(conn, name),
+		Timeout:    10 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("[WARN] Error waiting for JobQueue state to be \"VALID\": %s", err)
+	}
+
+	arn := *out.JobQueueArn
+	log.Printf("[DEBUG] JobQueue created: %s", arn)
+	d.SetId(arn)
+
+	return resourceAwsBatchJobQueueRead(d, meta)
+}
+
+func resourceAwsBatchJobQueueRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+
+	jq, err := getJobQueue(conn, d.Get("name").(string))
+	if err != nil {
+		return err
+	}
+	if jq == nil {
+		return fmt.Errorf("[WARN] Error reading JobQueue: \"%s\"", err)
+	}
+	d.Set("arn", jq.JobQueueArn)
+	d.Set("compute_environments", jq.ComputeEnvironmentOrder)
+	d.Set("name", jq.JobQueueName)
+	d.Set("priority", jq.Priority)
+	d.Set("state", jq.State)
+	return nil
+}
+
+func resourceAwsBatchJobQueueUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+
+	updateInput := &batch.UpdateJobQueueInput{
+		ComputeEnvironmentOrder: createComputeEnvironmentOrder(d),
+		JobQueue:                aws.String(d.Get("name").(string)),
+		Priority:                aws.Int64(int64(d.Get("priority").(int))),
+		State:                   aws.String(d.Get("state").(string)),
+	}
+	_, err := conn.UpdateJobQueue(updateInput)
+	if err != nil {
+		return err
+	}
+	return resourceAwsBatchJobQueueRead(d, meta)
+}
+
+func resourceAwsBatchJobQueueDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).batchconn
+	sn := d.Get("name").(string)
+	_, err := conn.UpdateJobQueue(&batch.UpdateJobQueueInput{
+		JobQueue: aws.String(sn),
+		State:    aws.String(batch.JQStateDisabled),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Wait until the Job Queue is disabled
+	err = resource.Retry(5*time.Minute, func() *resource.RetryError {
+		log.Printf("[DEBUG] Trying to delete Job Queue %s", sn)
+		_, err = conn.DeleteJobQueue(&batch.DeleteJobQueueInput{
+			JobQueue: aws.String(sn),
+		})
+		if err == nil {
+			return nil
+		}
+		return resource.RetryableError(err)
+	})
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{batch.JQStatusUpdating, batch.JQStatusDeleting},
+		Target:     []string{batch.JQStatusDeleted},
+		Refresh:    batchJobQueueRefreshFunc(conn, sn),
+		Timeout:    10 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for Job Queue (%s) to be destroyed: %s",
+			sn, err)
+	}
+	d.SetId("")
+	return nil
+}
+
+func createComputeEnvironmentOrder(d *schema.ResourceData) (envs []*batch.ComputeEnvironmentOrder) {
```

Comment:
Maybe a nitpick, but the interface here seems to be quite greedy - any reason we can't just pass an `[]interface{}` and call `d.Get()` outside of this function since it only needs access to a single field, not all ResourceData?